engine/src/damage/damageable_layer.cpp

```diff
@@ -283,14 +348,57 @@ float DamageableLayer::GetPercent(FacetName facet_name) {
     return percent;
 }
 
-void DamageableLayer::Regenerate() {
+void DamageableLayer::Regenerate(float recharge_rate) {
     for(Health& facet : facets) {
-        facet.Regenerate();
+        facet.Regenerate(recharge_rate);
     }
 }
 
+
+void DamageableLayer::RegenerateOrDischarge(float recharge_rate, bool velocity_discharge, float discharge_rate) {
+    for(Health& facet : facets) {
+        facet.Regenerate(recharge_rate);
+    }
+}
+
+
+float DamageableLayer::GetRegeneration() {
+    if(number_of_facets == 0) {
+        return 0.0f;
+    }
+
+    return facets[0].regeneration;
+}
+
+void DamageableLayer::UpdateFacets(const unsigned int new_size, const float new_facets[4]) {
+    assert(new_size == number_of_facets);
+
+    switch (number_of_facets) {
+    case 1:
+        facets[0].health = facets[0].max_health = new_facets[0];
+        break;
+
+    case 4:
+        facets[0].health = facets[0].max_health = new_facets[3];
+        facets[1].health = facets[1].max_health = new_facets[2];
+        facets[2].health = facets[2].max_health = new_facets[0];
+        facets[3].health = facets[3].max_health = new_facets[1];
+        break;
+    case 2:
```

Comment:
So case 2 falls through to case 8? Might want to add a comment to that effect. (Small nitpick.)